### PR TITLE
Add type hints for mmcv/fileio

### PR DESCRIPTION
**Pull request by triple-Mu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add type hints for mmcv/fileio

## Modification

- Update mmcv/fileio

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/fileio/io.py [hunk 1]
```diff
@@ -43,15 +47,16 @@ def load(file, file_format=None, file_client_args=None, **kwargs):
     Returns:
         The content from the file.
     """
+    f: Union[StringIO, BytesIO]
```
Comment (zhouzaida) [inline mmcv/fileio/io.py:50]:
```suggestion
```

File: mmcv/fileio/io.py [hunk 2]
```diff
@@ -1,8 +1,9 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 from io import BytesIO, StringIO
 from pathlib import Path
+from typing import Any, Callable, Dict, Optional, Union
 
-from ..utils import is_list_of, is_str
+from ..utils import is_list_of
 from .file_client import FileClient
 from .handlers import BaseFileHandler, JsonHandler, PickleHandler, YamlHandler
```
Comment (zhouzaida) [inline mmcv/fileio/io.py:8]:
```suggestion
from ..utils import is_list_of
from .file_client import FileClient
from .handlers import BaseFileHandler, JsonHandler, PickleHandler, YamlHandler

FileLikeObject = Union[StringIO, BytesIO]
```

File: mmcv/fileio/io.py [hunk 3]
```diff
@@ -43,15 +47,16 @@ def load(file, file_format=None, file_client_args=None, **kwargs):
...
         file = str(file)
-    if file_format is None and is_str(file):
+    if file_format is None and isinstance(file, str):
         file_format = file.split('.')[-1]
     if file_format not in file_handlers:
         raise TypeError(f'Unsupported format: {file_format}')
 
     handler = file_handlers[file_format]
-    if is_str(file):
+    if isinstance(file, str):
```
Comment (zhouzaida) [inline mmcv/fileio/io.py:59]:
```suggestion
    f: FileLikeObject
    if isinstance(file, str):
```

File: mmcv/fileio/io.py [hunk 4]
```diff
@@ -66,7 +71,11 @@ def load(file, file_format=None, file_client_args=None, **kwargs):
     return obj
 
 
-def dump(obj, file=None, file_format=None, file_client_args=None, **kwargs):
+def dump(obj: Any,
+         file: Optional[Union[str, Path]] = None,
```
Comment (zhouzaida) [inline mmcv/fileio/io.py:75]:
```suggestion
         file: Optional[Union[str, Path, FileLikeObject]] = None,
```

File: mmcv/fileio/io.py [hunk 5]
```diff
@@ -93,10 +102,11 @@ def dump(obj, file=None, file_format=None, file_client_args=None, **kwargs):
     Returns:
         bool: True for success, False otherwise.
     """
+    f: Union[StringIO, BytesIO]
```
Comment (zhouzaida) [inline mmcv/fileio/io.py:105]:
```suggestion
```

Comment (zhouzaida) [reply mmcv/fileio/io.py:105]: add `f: FileLikeObject` at line 117.

File: mmcv/fileio/io.py [hunk 6]
```diff
@@ -1,8 +1,9 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 from io import BytesIO, StringIO
 from pathlib import Path
+from typing import Any, Callable, Dict, Optional, Union
```
Comment (zhouzaida) [inline mmcv/fileio/io.py:4]:
```suggestion
from typing import Any, Callable, Dict, Optional, Union, List
```

File: mmcv/fileio/parse.py [hunk 7]
```diff
@@ -1,16 +1,17 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 
 from io import StringIO
+from typing import Dict, List, Optional
```
Comment (zhouzaida) [inline mmcv/fileio/parse.py:4]:
```suggestion
from pathlib import Path
from typing import Dict, List, Optional
```

File: mmcv/fileio/parse.py [hunk 8]
```diff
@@ -1,16 +1,17 @@
...
 from .file_client import FileClient
 
 
-def list_from_file(filename,
-                   prefix='',
-                   offset=0,
-                   max_num=0,
-                   encoding='utf-8',
-                   file_client_args=None):
+def list_from_file(filename: str,
```
Comment (zhouzaida) [inline mmcv/fileio/parse.py:9]:
```suggestion
def list_from_file(filename: Union[str, Path],
```

File: mmcv/fileio/parse.py [hunk 9]
```diff
@@ -52,10 +53,10 @@ def list_from_file(filename,
     return item_list
 
 
-def dict_from_file(filename,
-                   key_type=str,
-                   encoding='utf-8',
-                   file_client_args=None):
+def dict_from_file(filename: str,
```
Comment (zhouzaida) [inline mmcv/fileio/parse.py:56]:
```suggestion
def dict_from_file(filename: Union[str, Path],
```

**Comment by zhouzaida** [comment]
Hi, lint failed.

**Comment by triple-Mu** [comment]
> Hi, lint failed.

Got it. Fix now.

**Review by zhouzaida — APPROVED** [review]

**Review by imabackstabber — APPROVED** [review]
LGTM

**Pull request merged**